packages/core/src/node/messaging/ipc-connection-provider.ts

```diff
@@ -0,0 +1,107 @@
+    readonly errorHandler?: ConnectionErrorHandler
+}
+export type IPCConnectionOptions = Partial<ResolvedIPCConnectionOptions> & {
+    readonly serverName: string
+    readonly entryPoint: string
+};
+
+@injectable()
+export class IPCConnectionProvider {
+
+    @inject(ILogger)
+    protected readonly logger: ILogger;
+
+    listen(options: IPCConnectionOptions, acceptor: (connection: MessageConnection) => void): Disposable {
+        return this.doListen({
+            logger: this.logger,
+            ...options
+        }, acceptor);
+    }
+
+    protected doListen(options: ResolvedIPCConnectionOptions, acceptor: (connection: MessageConnection) => void): Disposable {
+        const childProcess = this.fork(options);
+        const connection = this.createConnection(childProcess, options);
+        const toStop = new DisposableCollection();
+        const toCancelStop = toStop.push(Disposable.create(() => childProcess.kill()));
+        const errorHandler = options.errorHandler;
+        if (errorHandler) {
+            connection.onError((e: [Error, Message | undefined, number | undefined]) => {
+                if (errorHandler.shouldStop(e[0], e[1], e[2])) {
+                    toStop.dispose();
+                }
+            });
+            connection.onClose(() => {
+                if (toStop.disposed) {
+                    return;
+                }
+                const error = errorHandler.shouldRestart();
+                if (error) {
+                    this.logger.error(error);
+                } else {
+                    toCancelStop.dispose();
+                    toStop.push(this.doListen(options, acceptor));
+                }
+            });
+        }
+        acceptor(connection);
+        return toStop;
+    }
+
+    protected createConnection(childProcess: cp.ChildProcess, options: ResolvedIPCConnectionOptions): MessageConnection {
+        const reader = new IPCMessageReader(childProcess);
+        const writer = new IPCMessageWriter(childProcess);
+        const connection = createMessageConnection(reader, writer, {
+            error: (message: string) => this.logger.error(`[${options.serverName}: ${childProcess.pid}] ${message}`),
+            warn: (message: string) => this.logger.warn(`[${options.serverName}: ${childProcess.pid}] ${message}`),
+            info: (message: string) => this.logger.info(`[${options.serverName}: ${childProcess.pid}] ${message}`),
+            log: (message: string) => this.logger.info(`[${options.serverName}: ${childProcess.pid}] ${message}`)
+        });
+        connection.trace(Trace.Off, {
+            log: (message, data) => this.logger.info(`[${options.serverName}: ${childProcess.pid}] ${message} ${data}`)
+        });
+        return connection;
+    }
+
+    protected fork(options: ResolvedIPCConnectionOptions): cp.ChildProcess {
+        const forkOptions: cp.ForkOptions = {
+            silent: true,
+            env: {
+                ...process.env
+            },
```

Comment:
Just asking to squeeze some additional knowledge from you: it is the same as `env: process.env`, right?

Reply:
no, it copies properties to a new object, later 2 more properties added to the new object but not to `process.env`